File: Introductiion To Recursion/Rope Cutting Problem/solutiondynamic.js
// Function to implement the dynamic programming rope cutting algorithm
function dynamicRopeCut(ropeLength, desiredLengths) {
    // Initialize an array to store minimum cuts for each length
    let minCuts = Array(ropeLength + 1).fill(Infinity);
    minCuts[0] = 0;

    // Iterate through rope lengths
    for (let i = 1; i <= ropeLength; i++) {
        // Iterate through desired lengths
        for (let length of desiredLengths) {
            // Check if the current length can be cut from the rope
            if (i >= length) {
                // Update the minimum cuts for the current length
                minCuts[i] = Math.min(minCuts[i], minCuts[i - length] + 1);
            }
        }
    }

    // Return the minimum cuts needed for the entire rope length
    return minCuts[ropeLength];
}

// Example usage
let ropeLength = 10;
let desiredLengths = [3, 5, 7];
let minCuts = dynamicRopeCut(ropeLength, desiredLengths);
console.log("Minimum cuts using dynamic programming:", minCuts);

File: Introductiion To Recursion/Rope Cutting Problem/solutiondynamic.py
# Function to implement the dynamic programming rope cutting algorithm
def dynamic_rope_cut(rope_length, desired_lengths):
    # Initialize an array to store minimum cuts for each length
    min_cuts = [float('inf')] * (rope_length + 1)
    min_cuts[0] = 0

    # Iterate through rope lengths
    for i in range(1, rope_length + 1):
        # Iterate through desired lengths
        for length in desired_lengths:
            # Check if the current length can be cut from the rope
            if i >= length:
                # Update the minimum cuts for the current length
                min_cuts[i] = min(min_cuts[i], min_cuts[i - length] + 1)

    # Return the minimum cuts needed for the entire rope length
    return min_cuts[rope_length]

# Example usage
rope_length = 10
desired_lengths = [3, 5, 7]
min_cuts = dynamic_rope_cut(rope_length, desired_lengths)
print("Minimum cuts using dynamic programming:", min_cuts)
